File: TP2-POC/src/tp2/poc/Cliente.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package tp2.poc;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.net.Socket;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 *
 * @author 123437
 */
public class Cliente {
    
    private String ip;
    private int porta;
    private Socket socket;
    
    public Cliente(String ip, int porta){
        this.ip = ip;
        this.porta = porta;
    }
    
    public void conectar(){
        try{
            socket = new Socket(ip, porta);
            System.out.println("Conectou ao servidor: "+ip+":"+porta);
        }
        catch(IOException ex){
            System.out.println("Falha ao conectar com servidor.");
            ex.printStackTrace();
            socket = null;
        }
        
    }
    
    public void enviar(Mensagem m){
        ObjectOutputStream saida = null;
        try {
            //while(socket == null);
            System.out.print("Cliente enviando "+m.toString()+"...");
            
            saida = new ObjectOutputStream(socket.getOutputStream());
            
            //saida.writeUTF(m.toString());
            saida.writeObject(m);
            
            saida.flush();
            System.out.println("Ok");

        } catch (IOException ex) {
            System.out.println("Falha ao enviar mensagem ao servidor.");
            ex.printStackTrace();
        }  /*finally {
            try {
                saida.close();
            } catch (IOException ex) {
                //System.out.println("Falha ao encerrar conexão de saída do cliente.");
                //ex.printStackTrace();
            }
        }*/
    }
    
    public Mensagem receber() {
        
        Mensagem m = null;
        ObjectInputStream entrada = null;
        try {
            while(socket == null);
            
            System.out.print("Cliente recebendo...");
            entrada = new ObjectInputStream(socket.getInputStream());
            
            //String mensagem = entrada.readUTF();
            //m = new Mensagem(mensagem);
            m = (Mensagem) entrada.readObject();
            
            System.out.println("Ok");

        } catch (IOException ex) {
            System.out.println("Falha ao receber mensagem do cliente.");
            ex.printStackTrace();
        } catch (ClassNotFoundException ex) { 
            Logger.getLogger(Cliente.class.getName()).log(Level.SEVERE, null, ex);
        }
        /*finally{
            try {
                entrada.close();
            } catch (IOException ex) {
                //System.out.println("Falha ao encerrar conexão de entrada do cliente.");
                //ex.printStackTrace();
            }
        }*/
        return m;
    }
    
}

File: TP3_Novo/P2P/Cliente.cpp
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */

/* 
 * File:   Cliente.cpp
 * Author: gustavo
 * 
 * Created on October 5, 2017, 9:27 PM
 */

#include "Cliente.h"

bool Cliente::sendingBusy = false;

Cliente::~Cliente() {
}

Cliente::Cliente(string ip, int porta) {
    this->ip = ip;
    this->porta = porta;
}

int Cliente::conectar() {
    struct sockaddr_in serv_addr;
    socklen_t * addrlen;

    serv_addr.sin_family = AF_INET;
    serv_addr.sin_addr.s_addr = inet_addr(ip.c_str());
    serv_addr.sin_port = htons(porta);

    conexao = socket(AF_INET, SOCK_STREAM, 0);
    if (connect(conexao, (struct sockaddr *) &serv_addr, sizeof (serv_addr)) < 0) {
        perror("Ocorreu um erro na conexão - connect");
        //exit(1);
        return -1;
    }
    
    else{
        printf("Conectou(%s:%d)\n", ip.c_str(), porta);
        return 1;
    }
    
}

int Cliente::conectarAposFalha() {
    struct sockaddr_in serv_addr;
    socklen_t * addrlen;

    serv_addr.sin_family = AF_INET;
    serv_addr.sin_addr.s_addr = inet_addr(ip.c_str());
    serv_addr.sin_port = htons(porta);

    conexao = socket(AF_INET, SOCK_STREAM, 0);
    // Retorna negativo quando não conseguiu conectar
    return connect(conexao, (struct sockaddr *) &serv_addr, sizeof (serv_addr));
}


void Cliente::desconectar() {
    close(conexao);
}

int Cliente::enviar(Mensagem* m) {
    sleep(1);
    while (sendingBusy);

    sendingBusy = true;

    /*
    printf("\nEnviando mensagem do cliente...");
    fflush(stdout);
     */

    int numBytes = 0;
    stringstream ss;
    ss << m->getCodigo();
    
    if(m->getTexto().compare("") != 0){
        ss << "|" << m->getTexto();
    }
    
    ss << "@";
    string str = ss.str();

    numBytes = ::send(conexao, str.c_str(), str.size(), 0);
    fsync(conexao);

    
    if (numBytes <= 0) {
        perror("\nNão foi possível enviar mensagem. - send");
        printf(" -> %s", str.c_str());
        fflush(stdout);
    } else {
        if(m->getCodigo() != Mensagem::PING){        
            printf("Enviou(%s)\n", str.c_str());
            fflush(stdout);
        }
    }
     

    sendingBusy = false;
    return numBytes;
}

Mensagem* Cliente::receber() {
    /*
    printf("\nRecebendo mensagem no cliente...");
    fflush(stdout);
     */
    Mensagem* m;

    int numBytes = 0;
    char msg[1000];
    numBytes = ::recv(conexao, msg, 1000, 0);

    if (numBytes > 0) {
        stringstream ss;
        ss << msg;
        string str = ss.str().substr(0, ss.str().find_first_of('@'));
        
        if(str.substr(0, 2).compare("12") != 0){
            printf("Recebeu(%s)\n", str.c_str());
            fflush(stdout);
        }
        return new Mensagem(str);
    } else {
        //perror("\nNão foi possível receber mensagem. - recv");
        return NULL;
    }
}

void Cliente::setEndereco(string ip, int porta) {
    this->ip = ip;
    this->porta = porta;
}

string Cliente::getIp() {
    return ip;
}

int Cliente::getPorta() {
    return porta;
}

bool Cliente::isSendingBusy() {
    return sendingBusy;
}
